internal/server/data/providergroup.go

```diff
@@ -0,0 +1,249 @@
+			return nil, err
+		}
+		result = append(result, pu)
+	}
+	return result, rows.Err()
+}
+
+type ListProviderGroupOptions struct {
+	ByProviderID       uid.ID
+	ByMemberIdentityID uid.ID
+}
+
+// ListProviderGroups returns all provider groups that match the specified criteria
+func ListProviderGroups(tx ReadTxn, opts ListProviderGroupOptions) ([]models.ProviderGroup, error) {
+	table := &providerGroupTable{}
+	query := querybuilder.New("SELECT")
+	query.B(columnsForSelect(table))
+	query.B("FROM")
+	query.B(table.Table())
+
+	if opts.ByMemberIdentityID != 0 {
+		query.B(`
+			JOIN provider_groups_provider_users 
+			ON provider_groups.name = provider_groups_provider_users.provider_group_name 
+			AND provider_groups.provider_id = provider_groups_provider_users.provider_id
+			`)
+	}
+
+	query.B("WHERE 1=1") // this is a condition which is always true, it allows us to append more conditions if needed
+	if opts.ByProviderID != 0 {
+		query.B("AND provider_groups.provider_id = ?", opts.ByProviderID)
+	}
+	if opts.ByMemberIdentityID != 0 {
+		query.B(`AND provider_groups_provider_users.provider_user_identity_id = ?`, opts.ByMemberIdentityID)
+	}
+
+	rows, err := tx.Query(query.String(), query.Args...)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var result []models.ProviderGroup
+	for rows.Next() {
+		var providerGroup models.ProviderGroup
+		fields := (*providerGroupTable)(&providerGroup).ScanFields()
+
+		if err := rows.Scan(fields...); err != nil {
+			return nil, err
+		}
+
+		result = append(result, providerGroup)
+	}
+
+	// load group members
+	for i := range result {
+		members, err := listProviderGroupMembers(tx, &result[i])
+		if err != nil {
+			return nil, fmt.Errorf("retrieve provider group members: %w", err)
+		}
+		result[i].Members = members
+	}
+	return result, rows.Err()
+}
+
+// setProviderGroupMembers ...
+func setProviderGroupMembers(tx GormTxn, providerGroup *models.ProviderGroup) error {
+	if len(providerGroup.Members) == 0 {
+		return nil // nothing to do
+	}
+	// the org does not need to be set here since all provider users and groups are org specific
+	stmt := `INSERT INTO provider_groups_provider_users ("provider_id", "provider_user_identity_id", "provider_group_name") VALUES`
+	values := []interface{}{}
+
+	val := 1
+	for _, user := range providerGroup.Members {
+		stmt += fmt.Sprintf("($%v, $%v, $%v),", val, val+1, val+2)
+		values = append(values, providerGroup.ProviderID, user.IdentityID, providerGroup.Name)
+		val += 3
+	}
+	// remove the trailing comma
+	stmt = strings.TrimSuffix(stmt, ",")
+	stmt += "ON CONFLICT DO NOTHING"
+
+	_, err := tx.Exec(stmt, values...)
+	if err != nil {
+		return fmt.Errorf("set provider group members %w", err)
+	}
+
+	return nil
+}
+
+// addMemberToProviderGroups adds a link between a provider user and group that exists in a provider
+func addMemberToProviderGroups(tx GormTxn, user *models.ProviderUser, providerGroupNames []string) error {
```

Comment:
stmt spacing might be nice here too. clearly not required to execute, but looks nicer in query logs.

Reply:
Im not sure what you mean by this one